### PR TITLE
Fix#41

### PR DESCRIPTION
**Pull request by mlindauer**
* extend api of random epm 
* add new facade
* add unit test to ROAR facade
* add hidden cmd line argument

### CODE REVIEW

File: examples/leadingones.py [hunk 1]
```diff
@@ -71,7 +71,6 @@ def leading_ones(cfg, seed):
 try:
     incumbent = smac.optimize()
 finally:
-    smac.stats.print_stats()
```
Comment (mfeurer) [inline examples/leadingones.py]: Why are these statements removed in this PR?

File: examples/rf.py [hunk 2]
```diff
@@ -122,7 +124,6 @@ def rfr(cfg, seed):
 try:
     incumbent = smac.optimize()
 finally:
-    smac.stats.print_stats()
```
Comment (mfeurer) [inline examples/rf.py]: Same here.

File: smac/epm/random_epm.py [hunk 3]
```diff
@@ -64,4 +64,42 @@ def predict(self, X):
         '''
         if not isinstance(X, np.ndarray):
             raise NotImplementedError("X has to be of type np.ndarray")
+        return self.rng.rand(len(X), 1), self.rng.rand(len(X), 1)
+    
+    def _predict(self, x):
```
Comment (mfeurer) [inline smac/epm/random_epm.py]: I think the inclusion of this method is a copy-paste error?

File: smac/facade/roar_facade.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+                 runhistory: RunHistory=None,
+                 intensifier: Intensifier=None,
+                 acquisition_function: AbstractAcquisitionFunction=None,
+                 model=None,
+                 runhistory2epm: AbstractRunHistory2EPM=None,
+                 initial_design: InitialDesign=None,
+                 stats: Stats=None,
+                 rng: np.random.RandomState=None):
+        '''
+        Facade to use ROAR default mode
```
Comment (mfeurer) [inline smac/facade/roar_facade.py]: What is the 'ROAR default mode'? I only know the 'ROAR' mode.

File: smac/facade/roar_facade.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+            run(self, config)
+            If not set, it will be initialized with the tae.ExecuteTARunOld()
+        runhistory: RunHistory
+            runhistory to store all algorithm runs
+        intensifier: Intensifier
+            intensification object to issue a racing to decide the current incumbent
+        acquisition_function : AcquisitionFunction
+            Object that implements the AbstractAcquisitionFunction. Will use
+            EI if not set.
+        model : object
```
Comment (mfeurer) [inline smac/facade/roar_facade.py]: I think the model shouldn't be an argument to the ROAR class

File: smac/facade/roar_facade.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+
+class ROAR(SMAC):
+
+    def __init__(self,
+                 scenario: Scenario,
+                 tae_runner: ExecuteTARun=None,
+                 runhistory: RunHistory=None,
+                 intensifier: Intensifier=None,
+                 acquisition_function: AbstractAcquisitionFunction=None,
+                 model=None,
```
Comment (mfeurer) [inline smac/facade/roar_facade.py]: I think the model shouldn't be an argument to the ROAR class.

File: smac/facade/roar_facade.py [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+
+
+class ROAR(SMAC):
+
+    def __init__(self,
+                 scenario: Scenario,
+                 tae_runner: ExecuteTARun=None,
+                 runhistory: RunHistory=None,
+                 intensifier: Intensifier=None,
+                 acquisition_function: AbstractAcquisitionFunction=None,
```
Comment (mfeurer) [inline smac/facade/roar_facade.py]: I don't think an acquisition function is necessary because we don't use a model.

File: smac/facade/roar_facade.py [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+        else:
+            raise TypeError('Unknown type %s for argument rng. Only accepts '
+                            'None, int or np.random.RandomState' % str(type(rng)))
+
+        # initial EPM
+        if model is None:
+            #use random predictions to simulate random sampling of configurations
+            model = RandomEpm(rng=rng)
+
+        # initial conversion of runhistory into EPM data
```
Comment (mfeurer) [inline smac/facade/roar_facade.py]: Why is this copied here? Can't this part of code be reused from the SMAC class?

File: smac/facade/roar_facade.py [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+        initial_design: InitialDesign
+            initial sampling design
+        stats: Stats
+            optional stats object
+        rng: np.random.RandomState
+            Random number generator
+        '''
+        self.logger = logging.getLogger("ROAR")
+
+        # initialize random number generator
```
Comment (mfeurer) [inline smac/facade/roar_facade.py]: This code duplication hints that setting the rng should be a private method in the SMAC class which can then be called in both constructors.

File: smac/facade/roar_facade.py [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+            num_run = rng
+            rng = np.random.RandomState(seed=rng)
+        elif isinstance(rng, np.random.RandomState):
+            num_run = rng.randint(1234567980)
+            rng = rng
+        else:
+            raise TypeError('Unknown type %s for argument rng. Only accepts '
+                            'None, int or np.random.RandomState' % str(type(rng)))
+
+        # initial EPM
```
Comment (mfeurer) [inline smac/facade/roar_facade.py]: In case ROAR accepts a model, we should in all cases replace it here an issue a warning.

File: smac/epm/random_epm.py (hunk 3 above)
Comment (mlindauer) [inline smac/epm/random_epm.py]: removed

File: examples/leadingones.py (hunk 1 above)
Comment (mlindauer) [inline examples/leadingones.py]: Because I wanted to merge a working PR without obvious bugs.

File: examples/rf.py (hunk 2 above)
Comment (mlindauer) [inline examples/rf.py]: same as above

File: smac/facade/roar_facade.py (hunk 7 above)
Comment (mlindauer) [inline smac/facade/roar_facade.py]: fixed

File: smac/facade/roar_facade.py (hunk 4 above)
Comment (mlindauer) [inline smac/facade/roar_facade.py]: fixed

File: smac/facade/roar_facade.py (hunk 5 above)
Comment (mlindauer) [inline smac/facade/roar_facade.py]: fixed

File: smac/facade/roar_facade.py (hunk 10 above)
Comment (mlindauer) [inline smac/facade/roar_facade.py]: obsolet

File: smac/facade/roar_facade.py (hunk 8 above)
Comment (mlindauer) [inline smac/facade/roar_facade.py]: because we can always disable the imputation of censored data.
We could remove the if-statement and always use RunHistory2EPM4Cost since the data is not used anyway.

File: examples/leadingones.py (hunk 1 above)
Comment (mfeurer) [inline examples/leadingones.py]: Sorry, I don't see the bug here

File: smac/facade/roar_facade.py (hunk 8 above)
Comment (mfeurer) [inline smac/facade/roar_facade.py]: If the data is not used anyway, wouldn't it be better to use a random imputer to reduce the running time of the ROAR algorithm?

Comment (mlindauer) [inline smac/facade/roar_facade.py]: right now, we use no imputer at all, which should be most time-efficient variant, right? We could replace the runhistory2EPM by a random class. However, this function is typically not the bottle neck. 
To save time, we would need to remove the local search and the sorting of the random configuration based on the acq function. However, this would require direct changes in the smbo class which I would like to avoid.

File: smac/facade/roar_facade.py (hunk 9 above)
Comment (mlindauer) [inline smac/facade/roar_facade.py]: but we need the rng for the random model!?

File: smac/facade/roar_facade.py (hunk 5 above)
Comment (KEggensperger) [inline smac/facade/roar_facade.py]: Is th docstring also already updated?

File: smac/facade/roar_facade.py (hunk 9 above)
Comment (KEggensperger) [inline smac/facade/roar_facade.py]: Agree, but maybe this should be a different PR

File: smac/facade/roar_facade.py (hunk 5 above)
Comment (mlindauer) [inline smac/facade/roar_facade.py]: fixed now.

**Comment by KEggensperger** [comment]
I guess we now want the base_epm code, as we use two different epms?

File: smac/facade/roar_facade.py (hunk 8 above)
Comment (mlindauer) [inline smac/facade/roar_facade.py]: Something has to be called by SMBO. We could add a nearly empty RunHistory2EPM class. However, more code could lead to more to maintain. I wanted to minimize the changes as much as possible

**Comment by mlindauer** [comment]
> I guess we now want the base_epm code, as we use two different epms?

to define the api? or why?

**Comment by KEggensperger** [comment]
Yes, but also to make the integration of new models simpler. Before we used only one epm so this was not necessary.

**Comment by mlindauer** [comment]
Should this be in the same PR or should we open a new one?

**Comment by mfeurer** [comment]
New PR please :)

**Comment by mlindauer** [comment]
Ok, are there any open issues for this PR or can we merge?

**Comment by mfeurer** [comment]
Here are the open todos:

- [ ] base class of EPM
- [ ] anwer my first question
- [ ] discuss the use of runhistory 2 EPM. I still find it weird that the user can pass this to ROAR. Couldn't we just remove the argument and pass `None` to the SMAC constructor?

File: examples/leadingones.py (hunk 1 above)
Comment (mlindauer) [inline examples/leadingones.py]: The optimize function of SMAC looks like

```
        try:
            incumbent = self.solver.run()
        finally:
            self.solver.stats.print_stats()
            self.logger.info("Final Incumbent: %s" % (self.solver.incumbent))
```

So, this would duplicate the output of the stats.

**Comment by mlindauer** [comment]
> base class of EPM

see commit

> anwer my first question

see answer

> discuss the use of runhistory 2 EPM. I still find it weird that the user can pass this to ROAR. Couldn't we just remove the argument and pass None to the SMAC constructor?

I agree that the argument in the ROAR facade is useless. Therefore, I removed it.
To avoid that SMAC will create a runhistory2epm with censored data imputation, I have to pass something to SMAC.

**Comment by mfeurer** [comment]
Awesome, we're getting close to be able to merge this. Could you please update the docstrings in the abstract EPM and change the author of the abstract EPM package? Then I also think this PR is ready to be merged.

**Comment by mlindauer** [comment]
> Could you please correct the docstrings in the abstract EPM 

done. I hope I haven't overlooked anything.

> and change the author of the abstract EPM package? 

funny to be the author for a bit of copy and pasta.
